### PR TITLE
Bitcoin exchange extravaganza

### PR DESCRIPTION
**Pull request by mrose17**
report `ledgerInfo.countryCode` and `ledgerInfo.exchangeInfo` for a new "Add Funds..." dialog.

auditor: @diracdeltas

### CODE REVIEW

File: docs/state.md [hunk 1]
```diff
@@ -435,8 +435,14 @@ WindowStore
       }
     },
     hasBitcoinHandler: boolean,  // brave browser has a `bitcoin:` URI handler
-    paymentIMG: string,           // the QR code equivalent of `paymentURL` expressed as "data:image/...;base64,..."
-    error: { // non-null if the last updateLedgerInfo happened concurrently with an error
+    countryCode:                 // ISO3166 2-letter code for country of browser's location
```
Comment (diracdeltas) [inline docs/state.md]: string

File: docs/state.md [hunk 2]
```diff
@@ -435,8 +435,14 @@ WindowStore
...
+    exchangeInfo: {              // information about corresponding "friendliest" BTC exchange (suggestions welcome!)
+      titleText: string,         // e.g., 'Create an account on ...'
+      buttonText: string,        // e.g., 'Create'
+      buttonURL: string,         // where the button should take the browser
```
Comment (diracdeltas) [inline docs/state.md]: will the title text and button text be localized?

File: app/ledger.js [hunk 3]
```diff
@@ -764,6 +772,27 @@ var updateLedgerInfo = () => {
...
+      if (err) console.log('ledger geoip warning: ' + JSON.stringify(err, null, 2))
+      if (result) ledgerInfo.countryCode = result
+
+      ledgerInfo.exchangeInfo = ledgerInfo._internal.exchanges[ledgerInfo.countryCode]
+
+      if (now <= ledgerInfo._internal.exchangeExpiry) return updateLedgerInfo()
+
+      ledgerInfo._internal.exchangeExpiry = now + msecs.day
+      roundtrip({ path: '/v1/exchange/providers' }, client.options, (err, response, body) => {
+        if (err) console.log('ledger exchange error: ' + JSON.stringify(err, null, 2))
```
Comment (diracdeltas) [inline app/ledger.js]: should this return if there's an error? or is this just a warning?

**Comment by diracdeltas** [comment]
please squash commits

File: app/ledger.js [hunk 4]
```diff
@@ -32,6 +32,7 @@ const session = electron.session
 const acorn = require('acorn')
 const ledgerBalance = require('ledger-balance')
 const ledgerClient = require('ledger-client')
+const ledgerGeoIP = require('ledger-geoip')
```
Comment (diracdeltas) [inline app/ledger.js]: please remove all providers in https://github.com/brave/ledger-geoip/blob/master/index.js that do not offer HTTPS

File: docs/state.md (hunk 2 above)
Comment (diracdeltas) [inline docs/state.md]: looks like https://ledger.brave.com/v1/exchange/providers is not localized, so take out 'buttonText' and 'titleText' and just add a property for the name of the exchange.

File: app/ledger.js [hunk 5]
```diff
@@ -764,6 +772,27 @@ var updateLedgerInfo = () => {
...
+      ledgerInfo.exchangeInfo = ledgerInfo._internal.exchanges[ledgerInfo.countryCode]
+
+      if (now <= ledgerInfo._internal.exchangeExpiry) return updateLedgerInfo()
+
+      ledgerInfo._internal.exchangeExpiry = now + msecs.day
+      roundtrip({ path: '/v1/exchange/providers' }, client.options, (err, response, body) => {
+        if (err) console.log('ledger exchange error: ' + JSON.stringify(err, null, 2))
+
+        ledgerInfo._internal.exchanges = body
+        ledgerInfo.exchangeInfo = ledgerInfo._internal.exchanges[ledgerInfo.countryCode]
```
Comment (diracdeltas) [inline app/ledger.js]: make sure `ledgerInfo._internal.exchanges` is not null/undefined

File: app/ledger.js (hunk 4 above)
Comment (mrose17) [inline app/ledger.js]: +1

File: app/ledger.js (hunk 3 above)
Comment (mrose17) [inline app/ledger.js]: it is actually, just a warning!

File: app/ledger.js (hunk 5 above)
Comment (mrose17) [inline app/ledger.js]: +1

File: docs/state.md (hunk 1 above)
Comment (mrose17) [inline docs/state.md]: +1

File: docs/state.md (hunk 2 above)
Comment (mrose17) [inline docs/state.md]: i suspect that localization has to occur in the browser in the ui subsystem. 

is there really a better key than just a text string? both `buttonText` and `titleText` can be used as the key values. am i missing something?

Comment (diracdeltas) [inline docs/state.md]: just send the name of the exchange, the front-end takes care of localization

**Comment by mrose17** [comment]
OBE'd by https://github.com/brave/browser-laptop/pull/3831

**Comment by alexwykoff** [comment]
Please add QA steps or label no-qa-needed as appropriate. Thanks!

**Comment by luixxiul** [comment]
bump for QA steps

**Comment by mrose17** [comment]
Steps to QA:

- use add funds from different CV countries and see what exchanges you see

**Pull request closed without merging**